flink-runtime/src/main/java/org/apache/flink/runtime/checkpoint/CheckpointRequestDecider.java

```diff
@@ -55,20 +55,23 @@
 	@GuardedBy("lock")
 	private final NavigableSet<CheckpointTriggerRequest> queuedRequests = new TreeSet<>(checkpointTriggerRequestsComparator());
 	private final int maxQueuedRequests;
+	private final Supplier<Long> completionTimeSupplier;
 
 	CheckpointRequestDecider(
 			int maxConcurrentCheckpointAttempts,
 			Consumer<Long> rescheduleTrigger,
 			Clock clock,
 			long minPauseBetweenCheckpoints,
 			Supplier<Integer> pendingCheckpointsSizeSupplier,
+			Supplier<Long> completionTimeSupplier,
```

Comment:
Those suppliers/consumers are making this class harder to follow and I think they are showing some deeper design problem - too tight coupling between `CheckpointCoordinator`/`CheckpointRequestDecider`.

In particular, it looks like adding `Supplier<Long> completionTimeSupplier` reveals quite weird concurrency contract between those two. Both are using the same lock object, and they are cross accessing the `CheckpointCoordinator#lastCheckpointCompletionRelativeTime` field from within the lock `CheckpointRequestDecider#lock` for read access and `CheckpointCoordinator#lock` for the write access.

What about making `CheckpointRequestDecider` a non thread safe class, remove the `CheckpointRequestDecider#lock` and just relay on the caller (`CheckpointCoordinator`) to either access `CheckpointRequestDecider` from a single thread or provide thread safety. For now `CheckpointCoordinator` would be acquiring `CheckpointCoordinator#lock` before calling `CheckpointRequestDecider`, while after completing threading model refactor, the `CheckpointCoordinator#lock` would be simply removed, without a need to modify `CheckpointRequestDecider`?

Reply:
I agree that the current design has these two issues:
1. Unnecessary complex
2. Unclear synchronization semantics

To deal with both of them I [proposed](https://docs.google.com/document/d/1p0m4FAmpWxShFaicgHXKqKCHYjnsDnKzvyU7BuA9CT8/edit?usp=sharing) to use actor-based approach.


OTH, the proposed change of "making CheckpointRequestDecider a non thread safe class":
1. replaces Suppliers with arguments, but doesn't remove the complexity
1. moves synchronization to another class, but doesn't remove it (we'll need to synchronize more sections in `CheckpointCoordinator`)

I like these changes, but not in the context of the upcoming refactoring (even if not actor-based - I tried to follow the original sync logic when extracting `CheckpointRequestDecider`, so changing it now can make the refactoring more difficult).

Besides that, the fix should be backported to 1.11, so I'd prefer a smaller changeset.

Reply:
As clarified offline, the proposal was not "remove existing suppliers" but "not to add a new one"; which is indeed a smaller changeset.
I'm updating the PR.